diff -Nru linux-2.6.36.orig/fs/squashfs/decompressor.c linux-2.6.36/fs/squashfs/decompressor.c
--- linux-2.6.36.orig/fs/squashfs/decompressor.c	2010-11-10 18:06:07.603046157 +0100
+++ linux-2.6.36/fs/squashfs/decompressor.c	2010-11-10 18:31:18.737955324 +0100
@@ -40,11 +40,9 @@
 	NULL, NULL, NULL, LZMA_COMPRESSION, "lzma", 0
 };
 
-#ifndef CONFIG_SQUASHFS_LZO
 static const struct squashfs_decompressor squashfs_lzo_unsupported_comp_ops = {
 	NULL, NULL, NULL, LZO_COMPRESSION, "lzo", 0
 };
-#endif
 
 static const struct squashfs_decompressor squashfs_unknown_comp_ops = {
 	NULL, NULL, NULL, 0, "unknown", 0
@@ -52,12 +50,12 @@
 
 static const struct squashfs_decompressor *decompressor[] = {
 	&squashfs_zlib_comp_ops,
-	&squashfs_lzma_unsupported_comp_ops,
-#ifdef CONFIG_SQUASHFS_LZO
-	&squashfs_lzo_comp_ops,
+#ifdef CONFIG_SQUASHFS_LZMA
+	&squashfs_lzma_comp_ops,
 #else
-	&squashfs_lzo_unsupported_comp_ops,
+	&squashfs_lzma_unsupported_comp_ops,
 #endif
+	&squashfs_lzo_unsupported_comp_ops,
 	&squashfs_unknown_comp_ops
 };
 
diff -Nru linux-2.6.36.orig/fs/squashfs/Kconfig linux-2.6.36/fs/squashfs/Kconfig
--- linux-2.6.36.orig/fs/squashfs/Kconfig	2010-11-10 18:06:07.604046138 +0100
+++ linux-2.6.36/fs/squashfs/Kconfig	2010-11-10 18:31:22.467902589 +0100
@@ -5,13 +5,13 @@
 	help
 	  Saying Y here includes support for SquashFS 4.0 (a Compressed
 	  Read-Only File System).  Squashfs is a highly compressed read-only
-	  filesystem for Linux.  It uses zlib/lzo compression to compress both
+	  filesystem for Linux.  It uses zlib compression to compress both
 	  files, inodes and directories.  Inodes in the system are very small
 	  and all blocks are packed to minimise data overhead. Block sizes
 	  greater than 4K are supported up to a maximum of 1 Mbytes (default
 	  block size 128K).  SquashFS 4.0 supports 64 bit filesystems and files
 	  (larger than 4GB), full uid/gid information, hard links and
-	  timestamps.
+	  timestamps.  
 
 	  Squashfs is intended for general read-only filesystem use, for
 	  archival use (i.e. in cases where a .tar.gz file may be used), and in
@@ -26,7 +26,7 @@
 
 	  If unsure, say N.
 
-config SQUASHFS_XATTR
+config SQUASHFS_XATTRS
 	bool "Squashfs XATTR support"
 	depends on SQUASHFS
 	default n
@@ -37,24 +37,15 @@
 
 	  If unsure, say N.
 
-config SQUASHFS_LZO
-	bool "Include support for LZO compressed file systems"
+config SQUASHFS_LZMA
+	bool "Include support for LZMA compressed file systems"
 	depends on SQUASHFS
-	default n
-	select LZO_DECOMPRESS
-	help
-	  Saying Y here includes support for reading Squashfs file systems
-	  compressed with LZO compresssion.  LZO compression is mainly
-	  aimed at embedded systems with slower CPUs where the overheads
-	  of zlib are too high.
-
-	  LZO is not the standard compression used in Squashfs and so most
-	  file systems will be readable without selecting this option.
-
-	  If unsure, say N.
+	select DECOMPRESS_LZMA
+	select DECOMPRESS_LZMA_NEEDED
 
 config SQUASHFS_EMBEDDED
-	bool "Additional option for memory-constrained systems"
+
+	bool "Additional option for memory-constrained systems" 
 	depends on SQUASHFS
 	default n
 	help
diff -Nru linux-2.6.36.orig/fs/squashfs/lzma_wrapper.c linux-2.6.36/fs/squashfs/lzma_wrapper.c
--- linux-2.6.36.orig/fs/squashfs/lzma_wrapper.c	1970-01-01 01:00:00.000000000 +0100
+++ linux-2.6.36/fs/squashfs/lzma_wrapper.c	2010-11-10 18:31:18.741955268 +0100
@@ -0,0 +1,152 @@
+/*
+ * Squashfs - a compressed read only filesystem for Linux
+ *
+ * Copyright (c) 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009
+ * Phillip Lougher <phillip@lougher.demon.co.uk>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ * lzma_wrapper.c
+ */
+
+#include <asm/unaligned.h>
+#include <linux/buffer_head.h>
+#include <linux/mutex.h>
+#include <linux/vmalloc.h>
+#include <linux/decompress/unlzma.h>
+#include <linux/slab.h>
+
+#include "squashfs_fs.h"
+#include "squashfs_fs_sb.h"
+#include "squashfs_fs_i.h"
+#include "squashfs.h"
+#include "decompressor.h"
+
+struct squashfs_lzma {
+	void	*input;
+	void	*output;
+};
+
+/* decompress_unlzma.c is currently non re-entrant... */
+DEFINE_MUTEX(lzma_mutex);
+
+/* decompress_unlzma.c doesn't provide any context in its callbacks... */
+static int lzma_error;
+
+static void error(char *m)
+{
+	ERROR("unlzma error: %s\n", m);
+	lzma_error = 1;
+}
+
+	
+static void *lzma_init(struct squashfs_sb_info *msblk)
+{
+	struct squashfs_lzma *stream = kzalloc(sizeof(*stream), GFP_KERNEL);
+	if (stream == NULL)
+		goto failed;
+	stream->input = vmalloc(msblk->block_size);
+	if (stream->input == NULL)
+		goto failed;
+	stream->output = vmalloc(msblk->block_size);
+	if (stream->output == NULL)
+		goto failed2;
+
+	return stream;
+
+failed2:
+	vfree(stream->input);
+failed:
+	ERROR("failed to allocate lzma workspace\n");
+	kfree(stream);
+	return NULL;
+}
+
+
+static void lzma_free(void *strm)
+{
+	struct squashfs_lzma *stream = strm;
+
+	if (stream) {
+		vfree(stream->input);
+		vfree(stream->output);
+	}
+	kfree(stream);
+}
+
+
+static int lzma_uncompress(struct squashfs_sb_info *msblk, void **buffer,
+	struct buffer_head **bh, int b, int offset, int length, int srclength,
+	int pages)
+{
+	struct squashfs_lzma *stream = msblk->stream;
+	void *buff = stream->input;
+	int avail, i, bytes = length, res;
+
+	mutex_lock(&lzma_mutex);
+
+	for (i = 0; i < b; i++) {
+		wait_on_buffer(bh[i]);
+		if (!buffer_uptodate(bh[i]))
+			goto block_release;
+
+		avail = min(bytes, msblk->devblksize - offset);
+		memcpy(buff, bh[i]->b_data + offset, avail);
+		buff += avail;
+		bytes -= avail;
+		offset = 0;
+		put_bh(bh[i]);
+	}
+
+	lzma_error = 0;
+	res = unlzma(stream->input, length, NULL, NULL, stream->output, NULL,
+							error);
+	if (res || lzma_error)
+		goto failed;
+
+	/* uncompressed size is stored in the LZMA header (5 byte offset) */
+	res = bytes = get_unaligned_le32(stream->input + 5);
+	for (i = 0, buff = stream->output; bytes && i < pages; i++) {
+		avail = min_t(int, bytes, PAGE_CACHE_SIZE);
+		memcpy(buffer[i], buff, avail);
+		buff += avail;
+		bytes -= avail;
+	}
+	if (bytes)
+		goto failed;
+
+	mutex_unlock(&lzma_mutex);
+	return res;
+
+block_release:
+	for (; i < b; i++)
+		put_bh(bh[i]);
+
+failed:
+	mutex_unlock(&lzma_mutex);
+
+	ERROR("lzma decompression failed, data probably corrupt\n");
+	return -EIO;
+}
+
+const struct squashfs_decompressor squashfs_lzma_comp_ops = {
+	.init = lzma_init,
+	.free = lzma_free,
+	.decompress = lzma_uncompress,
+	.id = LZMA_COMPRESSION,
+	.name = "lzma",
+	.supported = 1
+};
+
diff -Nru linux-2.6.36.orig/fs/squashfs/lzo_wrapper.c linux-2.6.36/fs/squashfs/lzo_wrapper.c
--- linux-2.6.36.orig/fs/squashfs/lzo_wrapper.c	2010-11-10 18:06:07.604046138 +0100
+++ linux-2.6.36/fs/squashfs/lzo_wrapper.c	1970-01-01 01:00:00.000000000 +0100
@@ -1,136 +0,0 @@
-/*
- * Squashfs - a compressed read only filesystem for Linux
- *
- * Copyright (c) 2010 LG Electronics
- * Chan Jeong <chan.jeong@lge.com>
- *
- * This program is free software; you can redistribute it and/or
- * modify it under the terms of the GNU General Public License
- * as published by the Free Software Foundation; either version 2,
- * or (at your option) any later version.
- *
- * This program is distributed in the hope that it will be useful,
- * but WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
- * GNU General Public License for more details.
- *
- * You should have received a copy of the GNU General Public License
- * along with this program; if not, write to the Free Software
- * Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
- *
- * lzo_wrapper.c
- */
-
-#include <linux/mutex.h>
-#include <linux/buffer_head.h>
-#include <linux/slab.h>
-#include <linux/vmalloc.h>
-#include <linux/lzo.h>
-
-#include "squashfs_fs.h"
-#include "squashfs_fs_sb.h"
-#include "squashfs_fs_i.h"
-#include "squashfs.h"
-#include "decompressor.h"
-
-struct squashfs_lzo {
-	void	*input;
-	void	*output;
-};
-
-static void *lzo_init(struct squashfs_sb_info *msblk)
-{
-	int block_size = max_t(int, msblk->block_size, SQUASHFS_METADATA_SIZE);
-
-	struct squashfs_lzo *stream = kzalloc(sizeof(*stream), GFP_KERNEL);
-	if (stream == NULL)
-		goto failed;
-	stream->input = vmalloc(block_size);
-	if (stream->input == NULL)
-		goto failed;
-	stream->output = vmalloc(block_size);
-	if (stream->output == NULL)
-		goto failed2;
-
-	return stream;
-
-failed2:
-	vfree(stream->input);
-failed:
-	ERROR("Failed to allocate lzo workspace\n");
-	kfree(stream);
-	return NULL;
-}
-
-
-static void lzo_free(void *strm)
-{
-	struct squashfs_lzo *stream = strm;
-
-	if (stream) {
-		vfree(stream->input);
-		vfree(stream->output);
-	}
-	kfree(stream);
-}
-
-
-static int lzo_uncompress(struct squashfs_sb_info *msblk, void **buffer,
-	struct buffer_head **bh, int b, int offset, int length, int srclength,
-	int pages)
-{
-	struct squashfs_lzo *stream = msblk->stream;
-	void *buff = stream->input;
-	int avail, i, bytes = length, res;
-	size_t out_len = srclength;
-
-	mutex_lock(&msblk->read_data_mutex);
-
-	for (i = 0; i < b; i++) {
-		wait_on_buffer(bh[i]);
-		if (!buffer_uptodate(bh[i]))
-			goto block_release;
-
-		avail = min(bytes, msblk->devblksize - offset);
-		memcpy(buff, bh[i]->b_data + offset, avail);
-		buff += avail;
-		bytes -= avail;
-		offset = 0;
-		put_bh(bh[i]);
-	}
-
-	res = lzo1x_decompress_safe(stream->input, (size_t)length,
-					stream->output, &out_len);
-	if (res != LZO_E_OK)
-		goto failed;
-
-	res = bytes = (int)out_len;
-	for (i = 0, buff = stream->output; bytes && i < pages; i++) {
-		avail = min_t(int, bytes, PAGE_CACHE_SIZE);
-		memcpy(buffer[i], buff, avail);
-		buff += avail;
-		bytes -= avail;
-	}
-
-	mutex_unlock(&msblk->read_data_mutex);
-	return res;
-
-block_release:
-	for (; i < b; i++)
-		put_bh(bh[i]);
-
-failed:
-	mutex_unlock(&msblk->read_data_mutex);
-
-	ERROR("lzo decompression failed, data probably corrupt\n");
-	return -EIO;
-}
-
-const struct squashfs_decompressor squashfs_lzo_comp_ops = {
-	.init = lzo_init,
-	.free = lzo_free,
-	.decompress = lzo_uncompress,
-	.id = LZO_COMPRESSION,
-	.name = "lzo",
-	.supported = 1
-};
diff -Nru linux-2.6.36.orig/fs/squashfs/Makefile linux-2.6.36/fs/squashfs/Makefile
--- linux-2.6.36.orig/fs/squashfs/Makefile	2010-11-10 18:06:07.603046157 +0100
+++ linux-2.6.36/fs/squashfs/Makefile	2010-11-10 18:31:18.736955339 +0100
@@ -5,5 +5,5 @@
 obj-$(CONFIG_SQUASHFS) += squashfs.o
 squashfs-y += block.o cache.o dir.o export.o file.o fragment.o id.o inode.o
 squashfs-y += namei.o super.o symlink.o zlib_wrapper.o decompressor.o
-squashfs-$(CONFIG_SQUASHFS_XATTR) += xattr.o xattr_id.o
-squashfs-$(CONFIG_SQUASHFS_LZO) += lzo_wrapper.o
+squashfs-$(CONFIG_SQUASHFS_LZMA) += lzma_wrapper.o
+
diff -Nru linux-2.6.36.orig/fs/squashfs/squashfs.h linux-2.6.36/fs/squashfs/squashfs.h
--- linux-2.6.36.orig/fs/squashfs/squashfs.h	2010-11-10 18:06:07.603046157 +0100
+++ linux-2.6.36/fs/squashfs/squashfs.h	2010-11-10 18:31:18.742955254 +0100
@@ -105,5 +105,5 @@
 /* zlib_wrapper.c */
 extern const struct squashfs_decompressor squashfs_zlib_comp_ops;
 
-/* lzo_wrapper.c */
-extern const struct squashfs_decompressor squashfs_lzo_comp_ops;
+/* lzma wrapper.c */
+extern const struct squashfs_decompressor squashfs_lzma_comp_ops;
diff -Nru linux-2.6.36.orig/fs/squashfs/xattr.h linux-2.6.36/fs/squashfs/xattr.h
--- linux-2.6.36.orig/fs/squashfs/xattr.h	2010-11-10 18:06:07.603046157 +0100
+++ linux-2.6.36/fs/squashfs/xattr.h	2010-11-10 18:31:15.721997889 +0100
@@ -21,7 +21,7 @@
  * xattr.h
  */
 
-#ifdef CONFIG_SQUASHFS_XATTR
+#ifdef CONFIG_SQUASHFS_XATTRS
 extern __le64 *squashfs_read_xattr_id_table(struct super_block *, u64,
 		u64 *, int *);
 extern int squashfs_xattr_lookup(struct super_block *, unsigned int, int *,
diff -Nru linux-2.6.36.orig/include/linux/decompress/bunzip2_mm.h linux-2.6.36/include/linux/decompress/bunzip2_mm.h
--- linux-2.6.36.orig/include/linux/decompress/bunzip2_mm.h	1970-01-01 01:00:00.000000000 +0100
+++ linux-2.6.36/include/linux/decompress/bunzip2_mm.h	2010-11-10 18:31:22.467902589 +0100
@@ -0,0 +1,12 @@
+#ifndef BUNZIP2_MM_H
+#define BUNZIP2_MM_H
+
+#ifdef STATIC
+/* Code active when included from pre-boot environment: */
+#define INIT
+#else
+/* Compile for initramfs/initrd code only */
+#define INIT __init
+#endif
+
+#endif
diff -Nru linux-2.6.36.orig/include/linux/decompress/inflate_mm.h linux-2.6.36/include/linux/decompress/inflate_mm.h
--- linux-2.6.36.orig/include/linux/decompress/inflate_mm.h	1970-01-01 01:00:00.000000000 +0100
+++ linux-2.6.36/include/linux/decompress/inflate_mm.h	2010-11-10 18:31:22.468902575 +0100
@@ -0,0 +1,12 @@
+#ifndef INFLATE_MM_H
+#define INFLATE_MM_H
+
+#ifdef STATIC
+/* Code active when included from pre-boot environment: */
+#define INIT
+#else
+/* Compile for initramfs/initrd code only */
+#define INIT __init
+#endif
+
+#endif
diff -Nru linux-2.6.36.orig/include/linux/decompress/mm.h linux-2.6.36/include/linux/decompress/mm.h
--- linux-2.6.36.orig/include/linux/decompress/mm.h	2010-11-10 18:05:56.997245575 +0100
+++ linux-2.6.36/include/linux/decompress/mm.h	2010-11-10 18:31:22.472902519 +0100
@@ -63,8 +63,6 @@
 
 #define set_error_fn(x)
 
-#define INIT
-
 #else /* STATIC */
 
 /* Code active when compiled standalone for use when loading ramdisk: */
@@ -87,7 +85,6 @@
 static void(*error)(char *m);
 #define set_error_fn(x) error = x;
 
-#define INIT __init
 #define STATIC
 
 #include <linux/init.h>
diff -Nru linux-2.6.36.orig/include/linux/decompress/unlzma_mm.h linux-2.6.36/include/linux/decompress/unlzma_mm.h
--- linux-2.6.36.orig/include/linux/decompress/unlzma_mm.h	1970-01-01 01:00:00.000000000 +0100
+++ linux-2.6.36/include/linux/decompress/unlzma_mm.h	2010-11-10 18:31:22.473902505 +0100
@@ -0,0 +1,20 @@
+#ifndef UNLZMA_MM_H
+#define UNLZMA_MM_H
+
+#ifdef STATIC
+
+/* Code active when included from pre-boot environment: */
+#define INIT
+
+#elif defined(CONFIG_DECOMPRESS_LZMA_NEEDED)
+
+/* Make it available to non initramfs/initrd code */
+#define INIT
+#include <linux/module.h>
+#else
+
+/* Compile for initramfs/initrd code only */
+#define INIT __init
+#endif
+
+#endif
diff -Nru linux-2.6.36.orig/lib/decompress_bunzip2.c linux-2.6.36/lib/decompress_bunzip2.c
--- linux-2.6.36.orig/lib/decompress_bunzip2.c	2010-11-10 18:05:56.923246964 +0100
+++ linux-2.6.36/lib/decompress_bunzip2.c	2010-11-10 18:31:22.479902419 +0100
@@ -52,6 +52,7 @@
 #include <linux/slab.h>
 #endif /* STATIC */
 
+#include <linux/decompress/bunzip2_mm.h>
 #include <linux/decompress/mm.h>
 
 #ifndef INT_MAX
diff -Nru linux-2.6.36.orig/lib/decompress_inflate.c linux-2.6.36/lib/decompress_inflate.c
--- linux-2.6.36.orig/lib/decompress_inflate.c	2010-11-10 18:05:56.923246964 +0100
+++ linux-2.6.36/lib/decompress_inflate.c	2010-11-10 18:31:22.480902405 +0100
@@ -23,6 +23,7 @@
 
 #endif /* STATIC */
 
+#include <linux/decompress/inflate_mm.h>
 #include <linux/decompress/mm.h>
 
 #define GZIP_IOBUF_SIZE (16*1024)
diff -Nru linux-2.6.36.orig/lib/decompress_unlzma.c linux-2.6.36/lib/decompress_unlzma.c
--- linux-2.6.36.orig/lib/decompress_unlzma.c	2010-11-10 18:05:56.924246945 +0100
+++ linux-2.6.36/lib/decompress_unlzma.c	2010-11-10 18:31:22.481902391 +0100
@@ -36,6 +36,7 @@
 #include <linux/slab.h>
 #endif /* STATIC */
 
+#include <linux/decompress/unlzma_mm.h>
 #include <linux/decompress/mm.h>
 
 #define	MIN(a, b) (((a) < (b)) ? (a) : (b))
@@ -531,7 +532,7 @@
 
 
 
-STATIC inline int INIT unlzma(unsigned char *buf, int in_len,
+STATIC int INIT unlzma(unsigned char *buf, int in_len,
 			      int(*fill)(void*, unsigned int),
 			      int(*flush)(void*, unsigned int),
 			      unsigned char *output,
@@ -664,4 +665,6 @@
 {
 	return unlzma(buf, in_len - 4, fill, flush, output, posp, error_fn);
 }
+#elif defined(CONFIG_DECOMPRESS_LZMA_NEEDED)
+EXPORT_SYMBOL(unlzma);
 #endif
diff -Nru linux-2.6.36.orig/lib/Kconfig linux-2.6.36/lib/Kconfig
--- linux-2.6.36.orig/lib/Kconfig	2010-11-10 18:05:56.925246926 +0100
+++ linux-2.6.36/lib/Kconfig	2010-11-10 18:31:22.476902463 +0100
@@ -124,6 +124,9 @@
 	select LZO_DECOMPRESS
 	tristate
 
+config DECOMPRESS_LZMA_NEEDED
+	 boolean
+
 #
 # Generic allocator support is selected if needed
 #
--- /dev/null
+++ b/include/linux/decompress/unlzo_mm.h
@@ -0,0 +1,10 @@
+#ifndef UNLZO_MM_H
+#define UNLZO_MM_H
+
+#ifdef STATIC
+#define INIT
+#else
+#define INIT __init
+#endif
+
+#endif
--- a/lib/decompress_unlzo.c
+++ b/lib/decompress_unlzo.c
@@ -39,6 +39,7 @@
 
 #include <linux/types.h>
 #include <linux/lzo.h>
+#include <linux/decompress/unlzo_mm.h>
 #include <linux/decompress/mm.h>
 
 #include <linux/compiler.h>
