--- autoconfig.orig	2009-08-08 21:02:35.000000000 +0000
+++ autoconfig	2009-08-09 20:59:35.150203562 +0000
@@ -7,6 +7,7 @@
 ALSA="yes"
 APM="no"
 BRLTTY="no"
+CPUFREQ="yes"
 COLDPLUG="yes"
 DETECT="yes"
 DHCP="yes"
@@ -42,6 +43,7 @@
 				ACPI="no"
 				ALSA="no"
 				APM="no"
+				CPUFREQ="no"
 				COLDPLUG="no"
 				DETECT="no"
 				DMRAID="no"
@@ -63,6 +65,9 @@
 			noapm)
 				APM="no"
 			;;
+			nocpufreq)
+				CPUFREQ="no"
+			;;
 			nocoldplug|nohotplug)
 				COLDPLUG="no"
 			;;
@@ -438,7 +443,7 @@
 		then
 			/etc/init.d/hotplug start
 		else
-			unpack_firmware
+			#unpack_firmware
 			[ -x /sbin/udevtrigger ] && /sbin/udevtrigger
 		fi
 		[ -e /etc/sysconfig ] || mkdir -p /etc/sysconfig
@@ -464,6 +469,8 @@
 		modprobe fan >/dev/null 2>&1
 		modprobe button >/dev/null 2>&1
 		modprobe battery >/dev/null 2>&1
+		modprobe sbs >/dev/null 2>&1
+		modprobe sbshc >/dev/null 2>&1
 		modprobe ac >/dev/null 2>&1
 		modprobe thermal >/dev/null 2>&1
 		modprobe video >/dev/null 2>&1
@@ -474,6 +481,24 @@
 		einfo "Not Loading ACPI support ..."
 	fi
 
+	if [ "${CPUFREQ}" = "yes" ]
+	then
+		modprobe acpi-cpufreq >/dev/null 2>&1
+		modprobe cpufreq-nforce2 >/dev/null 2>&1
+		modprobe e_powersaver >/dev/null 2>&1
+		modprobe gx-suspmod >/dev/null 2>&1
+		modprobe longhaul >/dev/null 2>&1
+		modprobe longrun >/dev/null 2>&1
+		modprobe p4-clockmod >/dev/null 2>&1
+		modprobe powernow-k6 >/dev/null 2>&1
+		modprobe powernow-k7 >/dev/null 2>&1
+		modprobe powernow-k8 >/dev/null 2>&1
+		modprobe speedstep-centrino >/dev/null 2>&1
+		modprobe speedstep-ich >/dev/null 2>&1
+		modprobe speedstep-lib >/dev/null 2>&1
+		modprobe speedstep-smi >/dev/null 2>&1
+	fi
+
 	if [ "${IDEDMA}" = "yes" ]
 	then
 		[ -x /etc/init.d/hdparm ] && /etc/init.d/hdparm start
@@ -488,19 +513,6 @@
 		ewarn "PCMCIA disabled via cmdline ..."
 	fi
 
-	if [ "${DHCP}" = "no" ]
-	then
-		sed -i -e '/^ifconfig_eth.*dhcp.*/ s/^/#/' \
-			-e '/^iface_eth.*dhcp.*/ s/^/#/' \
-			-e '/^config_eth.*dhcp.*/ s/^/#/' \
-			/etc/conf.d/net
-		for i in `seq 0 4`
-		do
-			echo "config_eth${i}=( \"none\" )" >> /etc/conf.d/net
-		done
-		ewarn "Skipping DHCP broadcast ..."
-	fi
-
 	# Read in what hwsetup has found
 	[ -f /etc/sysconfig/gentoo ] && source /etc/sysconfig/gentoo
 
@@ -531,52 +543,6 @@
 	fi
 
 	[ "${DETECT}" = "no" ] && DHCP="no"
-	[ "${DETECT}" = "yes" ] \
-		&& NETDEVICES="$(awk -F: '/eth.:|tr.:|ath.:|wlan.:/{print $1}' /proc/net/dev 2>/dev/null)"
-
-	if [ -n "${NETDEVICES}" ]
-	then
-		for nics in ${NETDEVICES}
-		do
-			if [ "${DHCP}" = "yes" ]
-			then
-				einfo "Network device ${HILITE}${nics}${NORMAL} detected, DHCP broadcasting for IP ..."
-				if [ -f /var/run/dhcpcd-${nics}.pid ]
-				then
-					if [ -z "$(/sbin/ifconfig ${nics} | grep 'inet addr')" ]
-					then
-						kill $(cat /var/run/dhcpcd-${nics}.pid)
-						sleep 2
-						dhcpcd -n -h $(hostname) ${nics}
-					fi
-				fi
-			fi
-		done
-		if [ "${NFS}" = "yes" ]
-		then
-			[ -x /etc/init.d/portmap ] && /etc/init.d/portmap start
-			[ -x /etc/init.d/nfsmount ] && /etc/init.d/nfsmount start
-		fi
-		if [ "${PASSWD}" = "no" ]
-		then
-			echo "root:${PASSWORD}" | chpasswd  > /dev/null 2>&1
-		else
-			/etc/init.d/pwgen start
-		fi
-		if [ "${SSHD}" = "yes" ]
-		then
-			# If we have passwd= on the command line, we do not run pwgen and we
-			# set the root password to PASSWORD.
-			if [ "${PASSWD}" = "yes" ]
-			then
-				ewarn "WARNING: You are starting sshd with a scrambled root password!!!"
-				ewarn "WARNING: You need to set a root password to be able to login remotely."
-			fi
-			[ -x /etc/init.d/sshd ] && /etc/init.d/sshd start
-		fi
-	else
-		ewarn "No Network device auto detected ..."
-	fi
 
 	if [ "${ALSA}" = "yes" ]
 	then
