--- /bin/bashlogin.orig	2009-08-08 21:02:35.000000000 +0000
+++ /bin/bashlogin	2009-08-10 18:29:35.116589647 +0000
@@ -1,7 +1,8 @@
 #!/bin/bash
+#
+# This is only installed by baselayout when USE=livecd to enable
+# serial consoles for some architectures
+#
+
+exec /bin/login root
 
-export HOME=/root
-cat /etc/motd 2>/dev/null
-cd /root
-[[ -e .bash_profile ]] && source .bash_profile
-exec -l /bin/bash -i
