--- /lib/rc/sh/init.sh.orig     2009-10-27 21:19:57.000000000 +0000
+++ /lib/rc/sh/init.sh  2009-10-28 13:10:22.000000000 +0000
@@ -68,4 +68,15 @@
 	eend $?
 fi

+[ -f "/sbin/livecd-functions.sh" ] && source /sbin/livecd-functions.sh
+[ -f "/sbin/livecd-functions.sh" ] && livecd_read_commandline
+
+if [ -f "/sbin/livecd-functions.sh" -a -n "${CDBOOT}" ]
+then
+	ebegin "Updating inittab"
+	livecd_fix_inittab
+	eend $?
+	/sbin/telinit q &>/dev/null
+fi
+
 . "$RC_LIBEXECDIR"/sh/init-common-post.sh
